Refactor Interpreter command execution

fmatthew5876 opened the pull request:
This PR modifies the interpreter to use `SaveEventExecState` and `SaveEventExecFrame` chunks so that it executes more accurately and allows correct saving and loading of LSD chunks while events are running.

Changes include
* Refactoring call stack to use frame array instead of child interpreters
* Fixing wait commands
* Fixing key input proc
* Fixing wait for all movement
* Other cleanups

Control flow refactors using `subcommand_path` chunks will come in the next PR.

~~Depends: #1717~~
~~Depends: #1739~~
~~Depends: https://github.com/EasyRPG/liblcf/pull/324~~

Fix: #757
Fix: #1777

Test Cases from #1707 addressed by this PR
- [x] 9
- [x] 10
- [x] 11
- [x] 12
- [x] 13 - commands work in rm2k3e mode
- [ ] 13 - commands crash otherwise (We currently print a warning)
- [x] 14
- [x] 15
- [x] 16
- [x] 17
- [x] 18

fmatthew5876 commented:
I'm cutting this one here and will do the control flow / subcommand_path stuff in the next PR.

Test cases related to this have been listed and need to be verified.

fmatthew5876 commented:
All test cases pass, except that for doing "ThisEvent" from common events in Player causes warnings instead of crashing. Personally I think this should be fixed, but that falls under #1720 and I don't want a huge debate about it holding up this PR.

This is ready for review and merge.

fdelapena commented:
Diff between PR1717 and PR1740: https://github.com/fmatthew5876/Player/compare/scene_fixes...fmatthew5876:event_state

Ghabry commented:
No idea where to put this but profiling the script heavy game smashbros indicated that 25% of the whole CPU time is consumed by "list = _list" (now: frame->commands = _list).
Is there any way to get rid of this copy?

fmatthew5876 commented:
> No idea where to put this but profiling the script heavy game smashbros indicated that 25% of the whole CPU time is consumed by "list = _list" (now: frame->commands = _list).
> Is there any way to get rid of this copy?

Great find! That makes sense for games constantly loading new events.

Could you write up a github issue. From a first glance, this seems non-trivial to me. Especially since `frame->commands` in the LSD data is literally a copy.

Since event code never changes, it should be possible to just use something akin to C++20 `std::span` to just point to the desired event buffer. That will require some surgery possibly at the liblcf level too.

So lets make an issue and plan it out. I think this is worthy performance optimization.

Ghabry commented:
This breaks Yume2kki completely.
After fast-forwarding through this stupid text only intro the start logo will loop forever.

fmatthew5876 commented:
Yume 2kki bug was fixed. It was a silly mistake the call stack refactor piece.

fmatthew5876 commented:
Rebased to master.

Retested wait commands and confirmed this PR is compatible with RPG_RT.

fmatthew5876 commented:
Peuteris Grey issue is fixed in #1779

Ghabry commented:
I see that you removed CommonEvent refreshing in ``Game_Map::Refresh``.
Because this is gone you could also remove (in another PR, rebase madness) all the ``RefreshType`` stuff and replace it with a ``bool needs_refresh`` as it was years ago.

This was a performance optimisation to ensure that only toggled switches refresh common events but as only map events are refreshed now this is not needed anymore. This also means that games with many common events will run much faster now because switch refreshing of common events is gone :+1:

Ghabry commented:
``struct event_info`` was completely removed. This was used to make debugging easier (for me): You could directly see which event was running in the debugger (and when using child interpreters the complete callstack).
Is there still a simple way to achive this with the new code?

Ghabry commented:
I found a regression in the game log.[in] ([Download](https://gabgab.xyz/yumenikki/Yume%20Nikki%20Fan%20Game%20Archive%20%28October%2022%2C%202018%29/Log.%5Bin%5D/Original/Ver.%201.3/Log.%5Bin%5D.rar))

Go through the intro, start the computer, select "Enhanced Reality". Once entered you can't leave the ER menu anymore. That "Please turn on ER headset" doesn't always appear and that cancelling sometimes closes the whole PC window are game bugs.

Ghabry commented:
This branch seems to have a significant performance regression.
The game Frozen Triggers (http://viprpg17kh.php.xdomain.jp/main/04_game.php?id=48 click ドウンロード and extract with unar) runs in the Release (!) build with 1 FPS.

The 0.6.1 Debug build runs the game with 25 FPS.

Ghabry commented:
I'm not a fan of all these code duplications:
```cpp
auto* frame = GetFrame();
assert(frame);
const auto& list = frame->commands;
```

Maybe add a ``GetCommands()`` function which returns the list and does all these checks.

As with other stuff I accept this as part of later PRs when rebasing is too ugly.

Ghabry commented on src/game_interpreter_battle.cpp:
```diff
@@ -91,6 +91,10 @@ bool Game_Interpreter_Battle::CommandCallCommonEvent(RPG::EventCommand const& co
 	}
 
+	if ((int)_state.stack.size() > call_stack_limit) {
+		Output::Error("Call Event limit has been exceeded");
```
Before the message was: ``Interpreter: Maximum callstack depth (100) exceeded``, the 100 could be mentioned somewhere.

fmatthew5876 replied:
Fixed

Ghabry commented on src/game_interpreter.cpp:
```diff
@@ -1139,19 +1126,20 @@ std::vector<Game_Actor*> Game_Interpreter::GetActors(int mode, int id) {
 		if (event_id == 0) {
 			// With no map parent
-			Output::Warning("Can't use ThisEvent in common event %d: Not called from a map event", event_info.x);
+			Output::Warning("Can't use ThisEvent in common event: Not called from a map event");
```
is there no way anymore to get the common event id for better debug output?

fmatthew5876 replied:
I took at a look at this and it's not easy. Mostly because we would have to augment the call stack to store the id for each stack frame. I could easily store the id of the running event but I wouldn't be able to fix it when I restore it.

This would need some refactor to not use LSD data directory but implement the call stack in a similar way to keyinput with translation to/from LSD data.

I agree we should have a good interpreter debugging mechanism, with first class logging and other things.

Can we make an issue and punt on that for now? Don't want to add another complex commit to this which will rebase hell everything.

Ghabry commented on src/game_interpreter.cpp:
```diff
@@ -103,38 +90,27 @@ bool Game_Interpreter::IsRunning() const {
-	CancelMenuCall();
-
-	if (main_flag && depth == 0) {
+	if (main_flag) {
```
> Refactor event call stack 

did you check that this is really not dependent on "depth == 0"?

Ghabry replied:
ah nvm, you don't create new interpreters anymore, so this looks fine.

Ghabry commented on src/game_interpreter.cpp:
```diff
@@ -2422,125 +2551,108 @@ bool Game_Interpreter::CommandPlayMemorizedBGM(RPG::EventCommand const& /* com *
-		Game_Variables.Set(var_id, result);
-		Game_Map::SetNeedRefresh(Game_Map::Refresh_Map);
-	}
+	return 0;
```
> Refactor keyinputproc to use LSD chunks properly 

In the old code when holding e.g. Shift + Num0 the result would be Num0, in the new code it would be shift. Is this correct?

Ghabry replied:
yeah this change is wrong, when I just create an event which changes the graphic based on Confirm or Shift in RPG\_RT it shows shift (7) when holding both, in Player is now confirm (5)

fmatthew5876 replied:
For multiple keys pressed, the logic seems to be greatest key value wins. I totally overlooked this situation. This was a really great find!

Ghabry commented:
Found another regression.
Create an event with two pages, page 1 is empty, page 2 is:

```
Change Faceset: Actor 1
Message: Bla
CallEvent: ThisEvent, Page 1
Message: Blab
```

Blab is now shown. I found this by accident, actually wanted to put Blab on Page1 xD

fmatthew5876 commented:
> I'm not a fan of all these code duplications:
> 
> ```c++
> auto* frame = GetFrame();
> assert(frame);
> const auto& list = frame->commands;
> ```
> 
> Maybe add a `GetCommands()` function which returns the list and does all these checks.
> 
> As with other stuff I accept this as part of later PRs when rebasing is too ugly.

I did this initially to try to minimize the amount of diffs. But I can come up with a better solution. I agree the asserts are ugly.

fmatthew5876 commented:
> Found another regression.
> Create an event with two pages, page 1 is empty, page 2 is:
> 
> ```
> Change Faceset: Actor 1
> Message: Bla
> CallEvent: ThisEvent, Page 1
> Message: Blab
> ```
> 
> Blab is **not** shown. I found this by accident, actually wanted to put Blab on Page1 xD
> 
> Edit:
> When Page1 contains a command Blab is shown.

I had a bug in how I wasn't correctly handling empty events in my new call event code. I added a fix which fixes this and Peuteris Grey issue reported by @fdelapena

fmatthew5876 commented:
> This branch seems to have a significant performance regression.
> The game Frozen Triggers (http://viprpg17kh.php.xdomain.jp/main/04_game.php?id=48 click ドウンロード and extract with unar) runs in the Release (!) build with 1 FPS.
> 
> The 0.6.1 Debug build runs the game with 25 FPS.

Are you sure?

I just tried this game with master in get 0 fps on linux.

fmatthew5876 commented:
> I found a regression in the game log.[in] ([Download](https://gabgab.xyz/yumenikki/Yume%20Nikki%20Fan%20Game%20Archive%20%28October%2022%2C%202018%29/Log.%5Bin%5D/Original/Ver.%201.3/Log.%5Bin%5D.rar))
> 
> Go through the intro, start the computer, select "Enhanced Reality". Once entered you can't leave the ER menu anymore. That "Please turn on ER headset" doesn't always appear and that cancelling sometimes closes the whole PC window are game bugs.

This one is fixed now.

I have addressed all review comments and game regressions. Also rebased to master.

Ghabry commented:
Well the stuff I found before my vacation is fixed, so is fine for me now.
Will just merge this one now to get this finally in. But all the other PRs have to wait until next Sunday when I'm back home.
